src/Components/test/E2ETest/Tests/ThreadingAppTest.cs

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.E2ETesting;
+using OpenQA.Selenium;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETest.Tests;
+
+public class ThreadingAppTest
+    : ServerTestBase<BlazorWasmTestAppFixture<ThreadingApp.Program>>, IDisposable
+{
+    public ThreadingAppTest(
+        BrowserFixture browserFixture,
+        BlazorWasmTestAppFixture<ThreadingApp.Program> serverFixture,
+        ITestOutputHelper output)
+        : base(browserFixture, serverFixture, output)
+    {
+    }
+
+    protected override void InitializeAsyncCore()
+    {
+        Navigate("/", noReload: true);
+        WaitUntilLoaded();
+    }
+
+    [Fact]
+    public void HasTitle()
+    {
+        Assert.Equal("Blazor standalone", Browser.Title);
+    }
+
+    [Fact]
+    public void HasHeading()
+    {
+        Assert.Equal("Hello, world!", Browser.Exists(By.TagName("h1")).Text);
+    }
+
+    [Fact]
+    public void NavMenuHighlightsCurrentLocation()
+    {
+        var activeNavLinksSelector = By.CssSelector(".sidebar a.active");
+        var mainHeaderSelector = By.TagName("h1");
+
+        // Verify we start at home, with the home link highlighted
+        Assert.Equal("Hello, world!", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Home", item.Text.Trim()));
+
+        // Click on the "counter" link
+        Browser.Exists(By.LinkText("Counter")).Click();
+
+        // Verify we're now on the counter page, with that nav link (only) highlighted
+        Assert.Equal("Counter", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Counter", item.Text.Trim()));
+
+        // Verify we can navigate back to home too
+        Browser.Exists(By.LinkText("Home")).Click();
+        Assert.Equal("Hello, world!", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Home", item.Text.Trim()));
+    }
+
+    [Fact]
+    public async Task HasCounterPage()
+    {
+        // Navigate to "Counter"
+        Browser.Exists(By.LinkText("Counter")).Click();
+        Assert.Equal("Counter", Browser.Exists(By.TagName("h1")).Text);
+
+        var countDisplayElement = Browser.Exists(By.CssSelector("h1 + p"));
```

Comment:
If we're going to check that the text changes below, we need to also check its initial value otherwise the test could pass falsely because someone rephrased the text in the test app, e.g.:

```suggestion
        var countDisplayElement = Browser.Exists(By.CssSelector("h1 + p"));
        Browser.Equal("Current count: 0", () => countDisplayElement.Text);
```

However you also need to change the test app so that it does not start the timer until the test explicitly takes a gesture to do so (e.g., by clicking a button), otherwise the test could be flaky since the timer might fire before this line.

I know it might seem like this is a bit demanding but we've had years of refining our techniques to make the E2E tests actually reliable and it does require following these kinds of patterns. The alternative is having tests that are each individually 99.99% reliable and in practice that's just not enough to keep the builds passing :)